generic ephemeral volume checks

pohly opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

When adding the ephemeral volume feature, the special case for
PersistentVolumeClaim volume sources was overlooked in three places:
- host path check in kubelet
- node limit check in the scheduler
- volume zone

Of these, volume zone is deprecated an doesn't need to be changed to support ephemeral volumes.

#### Special notes for your reviewer:

@msau42 pointed out the gap in nodevolumelimits on chat. The other two places where found with:
```
git grep -i '\(name\|pvc\).*=.*ClaimName'
```

#### Does this PR introduce a user-facing change?
```release-note
Generic ephemeral volumes were not considered properly by the the node limits scheduler filter and the kubelet hostpath check.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1698-generic-ephemeral-volumes
```

pohly committed:
consider ephemeral volumes for host path and node limits check

When adding the ephemeral volume feature, the special case for
PersistentVolumeClaim volume sources in kubelet's host path and node
limits checks was overlooked. An ephemeral volume source is another
way of referencing a claim and has to be treated the same way.

pohly commented:
I can also split up this PR into three different PRs, one per modified directory.

pohly commented:
/retest

damemi commented on pkg/scheduler/framework/plugins/nodevolumelimits/csi.go:
```diff
@@ -103,7 +106,7 @@ func (pl *CSILimits) Filter(ctx context.Context, _ *framework.CycleState, pod *v
 	attachedVolumes := make(map[string]string)
 	for _, existingPod := range nodeInfo.Pods {
-		if err := pl.filterAttachableVolumes(csiNode, existingPod.Pod.Spec.Volumes, existingPod.Pod.Namespace, attachedVolumes); err != nil {
+		if err := pl.filterAttachableVolumes(pod, csiNode, existingPod.Pod.Spec.Volumes, existingPod.Pod.Namespace, attachedVolumes); err != nil {
```
I think you want this to be `existingPod`, since on line 157 you set `pvcName = pod.Name + "-" + vol.Name`, where in this case `pod` is the pod being scheduled and `vol` is each existing pod volumes

pohly replied:
Good catch. I wonder why it didn't get caught by the unit tests.

Before I change it, would it make sense to remove the separate `volumes` and `namespace` parameters from `filterAttachableVolumes` and instead work only with the pod fields? It makes the function less flexible but simpler overall. I was undecided when adding the `pod` parameter.

pohly replied:
It wasn't caught because none of the unit tests use an existing pod which has a generic ephemeral volume. Do I really need to add also tests for that? :sob:

damemi commented on pkg/scheduler/framework/plugins/nodevolumelimits/csi.go:
```diff
@@ -134,13 +137,28 @@ func (pl *CSILimits) Filter(ctx context.Context, _ *framework.CycleState, pod *v
+		case vol.PersistentVolumeClaim != nil:
+			pvcName = vol.PersistentVolumeClaim.ClaimName
+		case vol.Ephemeral != nil:
+			if !utilfeature.DefaultFeatureGate.Enabled(features.GenericEphemeralVolume) {
```
we recently started trying to remove dependencies on calling `utilfeature.DefaultFeatureGate.Enabled()` directly in plugins, instead going with a pattern that shims in the featuregate flag setting as a parameter in the plugin registry. See https://github.com/kubernetes/kubernetes/pull/98583 and https://github.com/kubernetes/kubernetes/pull/98446 for examples

pohly replied:
I'll change to that.

damemi commented on pkg/scheduler/framework/plugins/nodevolumelimits/csi.go:
```diff
@@ -134,13 +137,28 @@ func (pl *CSILimits) Filter(ctx context.Context, _ *framework.CycleState, pod *v
+		case vol.Ephemeral != nil:
+			if !utilfeature.DefaultFeatureGate.Enabled(features.GenericEphemeralVolume) {
+				return fmt.Errorf(
+					"volume %s is a generic ephemeral volume, but that feature is disabled in kube-scheduler",
```
Should an error like this be logged at the scheduler level, or kubelet/api validation? Featuregates are for the whole cluster, so this isn't something that would just be disabled in kube-scheduler

pohly replied:
You can always have version or feature gate skew. Also, the pod might have been created when the feature was enabled for the entire cluster, then it got disabled and the pod remained in etcd.

It is debatable whether the limits check then should simply skip over the volume or complain. The former is closer to the behavior before the feature was introduced, but the pod is not going to be usable once the feature gets disabled and then a clear error message is better. We decided to do the latter (see commit d2cc70ee2caf628e942ed5d4c49b810401c08acc) so here I am doing the same.

damemi replied:
> the pod might have been created when the feature was enabled for the entire cluster, then it got disabled and the pod remained in etcd.

this might be worth considering, because I *think* we are generally supposed to be resilient to stored objects in etcd. This could break things if the feature gate is enabled then disabled right?

pohly replied:
The code is resilient against lingering Pod objects with the disabled volume type, that's why we have this check here. What this code here does is provide a clear error message instead of silently falling through and letting the Pod fail later.

It doesn't break the scheduler, only the Pod, but the Pod is already broken when the feature gets disabled.

pohly replied:
My comment was about new pods with ephemeral volumes. What I hadn't considered are old, running pods with ephemeral volumes. Those most be counted regardless of the feature gate. Throwing an error would affect all new pods, including those which don't use ephemeral volumes.

I added that distinction and the corresponding unit tests.

damemi commented on pkg/scheduler/framework/plugins/nodevolumelimits/csi.go:
```diff
@@ -153,6 +171,11 @@ func (pl *CSILimits) filterAttachableVolumes(
 		}
 
+		// ... the PVC for an ephemeral volume must be owned by the pod.
+		if ephemeral && !metav1.IsControlledBy(pvc, pod) {
```
this could be checked in the switch case, right?

pohly replied:
It needs the `pvc` which is not yet available inside the switch statement.

damemi replied:
Makes sense, thanks for clarifying

damemi commented on pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go:
```diff
@@ -267,17 +268,36 @@ func (pl *nonCSILimits) Filter(ctx context.Context, _ *framework.CycleState, pod
+		case vol.PersistentVolumeClaim != nil:
+			pvcName = vol.PersistentVolumeClaim.ClaimName
+		case vol.Ephemeral != nil:
+			if !utilfeature.DefaultFeatureGate.Enabled(features.GenericEphemeralVolume) {
```
same comments as above

damemi commented on pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go:
```diff
@@ -267,17 +268,36 @@ func (pl *nonCSILimits) Filter(ctx context.Context, _ *framework.CycleState, pod
+			continue
+		}
+		if pvcName == "" {
+			return fmt.Errorf("PersistentVolumeClaim had no name")
```
is it possible to create a pod with a pvc with no name?

pohly replied:
That if check was there already. I don't know whether this can happen in practice or whether it would be caught by validation. I prefer defensive programming, so I think it makes sense to have it.

damemi commented on pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go:
```diff
@@ -291,6 +311,11 @@ func (pl *nonCSILimits) filterVolumes(volumes []v1.Volume, namespace string, fil
 			}
 
+			// ... the PVC for an ephemeral volume must be owned by the pod.
+			if ephemeral && !metav1.IsControlledBy(pvc, pod) {
```
same as above, this could be moved to the switch case

pohly replied:
Same here - it needs `pvc`  :grin:

damemi commented on pkg/scheduler/framework/plugins/volumezone/volume_zone.go:
```diff
@@ -103,10 +106,25 @@ func (pl *VolumeZone) Filter(ctx context.Context, _ *framework.CycleState, pod *
+		case volume.PersistentVolumeClaim != nil:
+			pvcName = volume.PersistentVolumeClaim.ClaimName
+		case volume.Ephemeral != nil:
+			if !utilfeature.DefaultFeatureGate.Enabled(features.GenericEphemeralVolume) {
```
see featuregate pattern comment above

damemi reviewed (commented):
A couple comments on general code patterns, I don't know the full context of this featuregate so I will have to look more into it but this looks like the right approach overall

ehashman commented:
/triage accepted
/priority important-longterm

ehashman commented:
/cc @msau42 
/assign @damemi 
for scheduling

/assign @SergeyKanzhelev 
for node

pohly commented:
I rebased to resolve the (minor) conflict in volume_zone.go.

pohly commented:
/retest

pohly commented:
@damemi: ping... did I address your comments?

damemi commented:
I think this is pretty good from the scheduler side. Sorry about the delay. @Huang-Wei could you double check this for me? Just so there's more awareness of these changes among the scheduling sig

Huang-Wei commented on pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go:
```diff
@@ -267,17 +275,36 @@ func (pl *nonCSILimits) Filter(ctx context.Context, _ *framework.CycleState, pod
+		}
+		if pvcName == "" {
+			return fmt.Errorf("PersistentVolumeClaim had no name")
+		} else {
```
nit: remove this `else` to save one level of indention.

pohly replied:
I didn't do that because I wanted to preserve the original indention of the code that I didn't modify. But I can change that... done.

Huang-Wei reviewed (commented):
Some nits. LGTM otherwise.

msau42 commented on pkg/scheduler/framework/plugins/nodevolumelimits/csi.go:
```diff
@@ -134,13 +138,28 @@ func (pl *CSILimits) Filter(ctx context.Context, _ *framework.CycleState, pod *v
+				)
+			}
+			// Generic ephemeral inline volumes also use a PVC,
+			// just with a computed name, and...
```
and what? If this comment is continuing on L175 I would just say that here and say  the second part is checked later. If we end up adding more logic in between these lines, then this comment will be very confusing to follow.

pohly replied:
These two indeed belong together. I've made the clearer.

msau42 replied:
Has this been updated?

pohly replied:
I'm sure I did, but I must have lost that change again during a rebase, or I never properly committed it - sorry!

One more time with feeling!

pohly replied:
I had only updated the comment in `non_csi.go`. Having two very similar files is annoying, but as I said in my other comment, I'd prefer to not turn this PR into a refactoring exercise of this filter.

msau42 commented on pkg/scheduler/framework/plugins/nodevolumelimits/csi_test.go:
```diff
@@ -443,6 +487,50 @@ func TestCSILimits(t *testing.T) {
+			ephemeralEnabled: true,
+			extraClaims:      []v1.PersistentVolumeClaim{*ephemeralClaim},
+			driverNames:      []string{ebsCSIDriverName},
+			existingPods:     []*v1.Pod{runningPod, csiEBSTwoVolPod},
```
Can we also add test cases where:
1) existing pods were ephemeral
2) existing pods were a mix of ephemeral + persistent

pohly replied:
Added. I also added the case for existing pods with the feature gate disabled, because that was a case that I hadn't handled properly before.

msau42 commented on pkg/scheduler/framework/plugins/volumezone/volume_zone.go:
```diff
@@ -104,10 +108,25 @@ func (pl *VolumeZone) Filter(ctx context.Context, _ *framework.CycleState, pod *
+		pvcName := ""
+		ephemeral := false
+		switch {
+		case volume.PersistentVolumeClaim != nil:
```
Is there a utility function we can make for this? It seems like this same logic is used in many places now.

pohly replied:
We discussed that when originally adding generic ephemeral volumes. The conclusion at that time was that there is no good place for such an utility function. It would be very small, too, so changing dependencies just for it might not be justified.

Do you have a suggestion where it could be place such that kube-controller-manager, kube-apiserver and kubelet can call it?

msau42 replied:
We have used pkg/volume/util in the past to put methods shared by scheduler and kubelet

pohly replied:
I have started defining and using such helper methods in https://github.com/pohly/kubernetes/commits/generic-ephemeral-volume-helpers

I'll also convert all other code, to be sure that the proposed function signature works in all cases. Then we can start the bike shedding about the function names :sweat_smile: 

But let's do that in a separate PR once this here is merged. I had to rebase once more because of the changes in `registry.go`.

msau42 commented on pkg/scheduler/framework/plugins/volumezone/volume_zone.go:
```diff
@@ -70,7 +74,7 @@ func (pl *VolumeZone) Name() string {
 // volume might allow region-wide access)
 //
-// Currently this is only supported with PersistentVolumeClaims, and looks to the labels
+// Currently this is only supported with PersistentVolumeClaims and generic ephemeral volumes, and looks to the labels
```
This predicate is deprecated in favor of volume topology now. We don't have to extend support for generic ephemeral volumes here.

pohly replied:
So shall I remove the entire commit which touches `volume_zone.go`?

pohly replied:
I've removed it and can bring it back if desired.

msau42 commented:
Should we also extend https://github.com/kubernetes/kubernetes/blob/17f3990ea1cdc1bd20c0d6e742c247b5b72beae0/test/e2e/storage/csi_mock_volume.go#L528?

pohly commented on pkg/scheduler/framework/plugins/nodevolumelimits/csi.go:
```diff
@@ -134,13 +138,28 @@ func (pl *CSILimits) Filter(ctx context.Context, _ *framework.CycleState, pod *v
 
 func (pl *CSILimits) filterAttachableVolumes(
-	csiNode *storagev1.CSINode, volumes []v1.Volume, namespace string, result map[string]string) error {
+	pod *v1.Pod, csiNode *storagev1.CSINode, volumes []v1.Volume, namespace string, result map[string]string) error {
```
I agree, this is useful. I had considered that earlier but then wasn't sure whether there might be cases where the volumes do not come from the pod.

Changed.

pohly commented on pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go:
```diff
@@ -267,17 +275,36 @@ func (pl *nonCSILimits) Filter(ctx context.Context, _ *framework.CycleState, pod
 }
 
-func (pl *nonCSILimits) filterVolumes(volumes []v1.Volume, namespace string, filteredVolumes sets.String) error {
+func (pl *nonCSILimits) filterVolumes(pod *v1.Pod, volumes []v1.Volume, namespace string, filteredVolumes sets.String) error {
```
Done.

pohly commented:
> Should we also extend
> 
> https://github.com/kubernetes/kubernetes/blob/17f3990ea1cdc1bd20c0d6e742c247b5b72beae0/test/e2e/storage/csi_mock_volume.go#L528
> ?

Yes, we can. I've added two more tests (one for existing ephemeral with new persistent, one for existing persistent with new ephemeral). They are slow, though.

pohly commented:
/retest

pohly commented:
Rebased to resolve a minor conflict in `pkg/scheduler/framework/plugins/registry.go`.

@msau42 : did my earlier changes address your comments?

pohly commented:
@Huang-Wei : is the PR okay now?

ahg-g commented on pkg/scheduler/framework/plugins/nodevolumelimits/csi.go:
```diff
@@ -134,25 +138,45 @@ func (pl *CSILimits) Filter(ctx context.Context, _ *framework.CycleState, pod *v
 
 		if err != nil {
-			klog.V(5).InfoS("Unable to look up PVC info", "PVC", fmt.Sprintf("%s/%s", namespace, pvcName))
+			klog.V(5).InfoS("Unable to look up PVC info", "PVC", fmt.Sprintf("%s/%s", pod.Namespace, pvcName))
```
If we can't look up the PVC now, shouldn't we return unschedulable since we can't determine whether or not scheduling the pod on the node will exceed the volume limit?

pohly replied:
No idea. I don't know this code well enough to have an opinion about it.

Can we move this and the refactoring suggestion into issues and have someone work on that later? My goal with this KEP wasn't to change the underlying logic of how this filter works.

ahg-g replied:
sure, but I would like to get an answer if we should actually followup on this or not.

@msau42 can you answer this one?

msau42 replied:
I think the concern here was for counting other pods that reference PVCs that don't exist. We don't want misconfigured other pods to block scheduling of the current pod. We probably could be more specific here and check for existence.

ahg-g replied:
We are using the same function for the incoming pod too when `newPod=true`. So should we return  unschedulable if this lookup fails and `newPod=true`?

pohly replied:
This question is not related to this PR, but rather the logic of this check without my modifications, right?

@msau42 : can you answer the question?

msau42 replied:
Yes, if we can't find the PVCs for our own pod, then we should return an error. If it's another pod, we should ignore or handle "does not exist" specifically.

Agree we can handle this as a separate bug fix since this logic was already existing.

msau42 replied:
thinking about this some more, @pohly does this become a bigger issue with ephemeral volumes since it's likely that a pvc may not exist immediately?

pohly replied:
I still don't understand all of the implications (= as in, what this function does and how it is called) to have an informed opinion.

From your older comment:
> Yes, if we can't find the PVCs for our own pod, then we should return an error. 

I've added that, regardless of `newPod`.

> If it's another pod, we should ignore or handle "does not exist" specifically.

With "it's another pod", do you mean the case where the PVC exists but is owned by another Pod? For the current Pod that's still the same problem: it needs a PVC and doesn't have it. Therefore I kept the "PVC is not owned by pod" error for that case.

pohly replied:
There were unit tests in `non_csi_test.go` that explicitly required that "missing PVC is not counted towards the PV limit". Did I understand correctly that you wanted me to change that so that missing PVCs are now errors?

I don't know what went into that original design decision.

ahg-g commented on pkg/scheduler/framework/plugins/nodevolumelimits/csi.go:
```diff
@@ -103,7 +107,7 @@ func (pl *CSILimits) Filter(ctx context.Context, _ *framework.CycleState, pod *v
 	attachedVolumes := make(map[string]string)
 	for _, existingPod := range nodeInfo.Pods {
-		if err := pl.filterAttachableVolumes(csiNode, existingPod.Pod.Spec.Volumes, existingPod.Pod.Namespace, attachedVolumes); err != nil {
+		if err := pl.filterAttachableVolumes(existingPod.Pod, csiNode, false /* existing pod */, attachedVolumes); err != nil {
```
@msau42 looking at this filter, we recompute the volume count on the node for each incoming pod, we should cache that in the scheduler's nodeInfo like we do for other resources (memory, cpu etc.).

also, it seems some the checks below are agnostic to the node (like checking the incoming pod PVC and computing its volume count); those can be factored out into a PreFilter instead of redundantly doing them for each node.

Huang-Wei replied:
> we should cache that in the scheduler's nodeInfo like we do for other resources (memory, cpu etc.).

+1 to have a struct caching the storage related attributes to reduce duplicated computation. It can be implemented in a follow-up, and better first aggregate the requirements from storage.

ahg-g replied:
created https://github.com/kubernetes/kubernetes/issues/101803

pohly commented:
Rebased because of a conflict in `pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go` with 447e55132a450f202004f5b058d9b5e3d2efb563.

pohly commented:
/retest

msau42 commented on pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go:
```diff
@@ -267,57 +275,84 @@ func (pl *nonCSILimits) Filter(ctx context.Context, _ *framework.CycleState, pod
-				klog.V(4).InfoS("Unable to look up PVC info, assuming PVC doesn't match predicate when counting limits", "PVC", fmt.Sprintf("%s/%s", namespace, pvcName), "err", err)
-				continue
+		pvcName := ""
+		ephemeral := false
```
How much of this logic is common between non_csi and csi versions that we can share?

pohly replied:
I'm not sure. I don't want introduce bugs when refactoring too much, so I'd like to punt on that.

pohly commented:
@ehashman: I think I have provided all the answers and made the changes that were asked for, so "Waiting on Author" is not quite the right state. We need confirmation from @msau42 and @ahg-g that the concerns that they raised can be addressed in separate PRs.

ahg-g commented:
> We need confirmation from @msau42 and @ahg-g that the concerns that they raised can be addressed in separate PRs.

I have an open question https://github.com/kubernetes/kubernetes/pull/100482#discussion_r636232673

msau42 commented:
/lgtm
/approve

There's a couple follow up items:
* refactor the retrieval of PVC into a util method
* fixing an existing bug when the pvc doesn't exist for the pod being scheduled

msau42 commented:
/retest

ahg-g commented:
/hold

please squash the commits

ahg-g commented:
/approve

pohly commented:
> please squash the commits

IMHO these are logically independent changes and thus better kept in different commits, but it's not worth arguing about, so I squashed.

ahg-g commented:
> > please squash the commits
> 
> IMHO these are logically independent changes and thus better kept in different commits, but it's not worth arguing about, so I squashed.

sg, I wouldn't argue about this either, I thought the commits were related to addressing comments and the best practice is to squash them. If not, then sure we can keep them as is.

msau42 commented:
/assign sjenning
for kubelet approval

msau42 commented:
/lgtm

ehashman commented:
/hold cancel
since commits were squashed https://github.com/kubernetes/kubernetes/pull/100482#issuecomment-852653855

ehashman commented:
/skip

pohly commented:
I had to rebase.

@ahg-g @msau42 : may I have your LGTM once more?

@sjenning: do you have time to look at this or should we pull in someone else for the kubelet part?

pohly commented:
/retest

Test flake.

msau42 commented:
/lgtm

msau42 reviewed (commented):
/hold
for one thought I had

pohly committed:
scheduler: fail the volume attach limit when PVC is missing

Previously, the situation was ignored, which might have had the effect that Pod
scheduling continued (?) even though the Pod+PVC weren't known to be in an
acceptable state.

pohly commented:
/retest

pohly commented:
/retest

pohly commented:
/retest

pohly commented:
/retest

pohly commented:
/retest

msau42 commented on pkg/scheduler/framework/plugins/nodevolumelimits/csi.go:
```diff
@@ -180,8 +180,9 @@ func (pl *CSILimits) filterAttachableVolumes(
 		if err != nil {
-			klog.V(5).InfoS("Unable to look up PVC info", "PVC", fmt.Sprintf("%s/%s", pod.Namespace, pvcName))
-			continue
+			// The PVC is required to proceed with Pod
```
There's two scenarios we're concerned about: 1) The PVC is missing for the pod being scheduled 2) The PVC is missing for a different Pod that has already been scheduled. 

For the first, we want to fail, but for the 2nd we want to ignore so that a pod from another namespace cannot cause a different pod to fail scheduling.

pohly replied:
So what you are saying is that we have two pods, one which can run and one which can't because of a missing PVC, returning an error for the second pod will cause scheduling of *both* pods to fail? Why is returning an error not limited to just one pod?

pohly replied:
Perhaps it'll be clearer if you give an example?

pohly replied:
We discussed this in a meeting and @msau42 explained that this function gets called in a loop for new and existing pods. The decision at some point was to ignore existing pods with missing volumes.

I've changed the error case so that it only triggers for new pods.

@msau42: please take another look

gnufied commented:
/lgtm

gnufied commented:
/assign
/lgtm cancel

pohly commented:
I had to rebase because the feature gate package imports in `pkg/kubelet/kubelet_pods_test.go` had been removed and I need them again.

ehashman commented:
@msau42 can the hold be removed?

msau42 commented on pkg/scheduler/framework/plugins/nodevolumelimits/csi.go:
```diff
@@ -180,6 +180,14 @@ func (pl *CSILimits) filterAttachableVolumes(
+				// The PVC is required to proceed with
+				// scheduling of a new pod because it cannot
+				// run without it. Bail out immediately.
+				return fmt.Errorf("look up PVC %s/%s: %v", pod.Namespace, pvcName, err)
```
should we provide a little more context to the error message, like "failed to ..."?

pohly replied:
I tend to write error messages as "<operation>: <error>". Then if it gets wrapped higher up, we don't accumulate those "failed to " comments. I've seen errors which were something like:
```
ERROR: failed to <do abc>: failed to <do xyz>: could not <do fgh>: original error here
```

That always seemed a bit redundant to me. But let me add it here.... done.

ahg-g replied:
We tend to say "looking up PVC ...." because it is clear this is an error from the context, so adding "failed to" is redundant.

pohly replied:
I like "looking up PVC" best, so I've switched to that.

@msau42 okay? I still need LGTM for this PR.

msau42 commented on pkg/scheduler/framework/plugins/nodevolumelimits/csi.go:
```diff
@@ -180,6 +180,14 @@ func (pl *CSILimits) filterAttachableVolumes(
+				return fmt.Errorf("look up PVC %s/%s: %v", pod.Namespace, pvcName, err)
+			}
+			// If the PVC is invalid, we don't count the volume because
+			// there's no guarantee that it belongs to the running predicate.
```
opened https://github.com/kubernetes/kubernetes/issues/105328 to followup on if we should ignore or count them.

msau42 commented:
/hold cancel

msau42 commented:
/lgtm

klueska reviewed (commented):
Small comment, but otherwise looks good from a sig-node / kubelet perspective.

klueska commented:
/lgtm
/approve
